{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/nfs/lambda_stor_01/data/apartin/projects/cepi/viral-segmatch/notebooks\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import json\n",
    "from pathlib import Path\n",
    "from time import time\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "# import seaborn as sns\n",
    "# import matplotlib.pyplot as plt\n",
    "\n",
    "# filepath = Path(__file__).parent\n",
    "filepath = Path(os.path.abspath(''))\n",
    "print(filepath)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/nfs/lambda_stor_01/data/apartin/projects/cepi/viral-segmatch/notebooks/../data/raw\n",
      "/nfs/lambda_stor_01/data/apartin/projects/cepi/viral-segmatch/notebooks/../data/raw/Bunya-from-datasets/Quality_GTOs\n",
      "/nfs/lambda_stor_01/data/apartin/projects/cepi/viral-segmatch/notebooks/../data/bunya_processed\n"
     ]
    }
   ],
   "source": [
    "# task_name = 'Bunya-from-datasets' # config!\n",
    "task_name = 'bunya_processed'\n",
    "data_dir = filepath / '../data'\n",
    "raw_data_dir = data_dir / 'raw'\n",
    "quality_gto_dir = raw_data_dir / 'Bunya-from-datasets/Quality_GTOs'\n",
    "\n",
    "output_dir = data_dir / task_name\n",
    "os.makedirs(output_dir, exist_ok=True)\n",
    "\n",
    "print(raw_data_dir)\n",
    "print(quality_gto_dir)\n",
    "print(output_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# contig_quality"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total files: 1572\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Contig</th>\n",
       "      <th>Segment</th>\n",
       "      <th>Copy_Num</th>\n",
       "      <th>Len</th>\n",
       "      <th>Min_Len</th>\n",
       "      <th>Max_len</th>\n",
       "      <th>AMB_Bases</th>\n",
       "      <th>Frac_AMB</th>\n",
       "      <th>Exceptions</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NC_086348.1</td>\n",
       "      <td>Large RNA Segment</td>\n",
       "      <td>1</td>\n",
       "      <td>6374</td>\n",
       "      <td>5764.0</td>\n",
       "      <td>10736.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NC_086347.1</td>\n",
       "      <td>Medium RNA Segment</td>\n",
       "      <td>1</td>\n",
       "      <td>4310</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NC_086346.1</td>\n",
       "      <td>Small RNA Segment</td>\n",
       "      <td>1</td>\n",
       "      <td>1863</td>\n",
       "      <td>1350.0</td>\n",
       "      <td>3435.0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        Contig             Segment  Copy_Num   Len  Min_Len  Max_len  \\\n",
       "0  NC_086348.1   Large RNA Segment         1  6374   5764.0  10736.0   \n",
       "1  NC_086347.1  Medium RNA Segment         1  4310      NaN      NaN   \n",
       "2  NC_086346.1   Small RNA Segment         1  1863   1350.0   3435.0   \n",
       "\n",
       "   AMB_Bases  Frac_AMB  Exceptions  \n",
       "0          0         0         NaN  \n",
       "1          0         0         NaN  \n",
       "2          0         0         NaN  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# contig_quality\n",
    "print(f\"Total files: {len(sorted(quality_gto_dir.glob('*.contig_quality')))}\")\n",
    "ex_file = 'GCF_031497195.1'\n",
    "cq = pd.read_csv(quality_gto_dir / f'{ex_file}.contig_quality', sep='\\t')\n",
    "cq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# feature_quality"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total files: 1572\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>Function</th>\n",
       "      <th>Evaluated</th>\n",
       "      <th>Prot_Len</th>\n",
       "      <th>Min_Len</th>\n",
       "      <th>Max_Len</th>\n",
       "      <th>Copy_Num</th>\n",
       "      <th>Exp_Copy_Num</th>\n",
       "      <th>Exceptions</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>fig|426786.38.CDS.1</td>\n",
       "      <td>Nucleocapsid protein (N protein)</td>\n",
       "      <td>1.0</td>\n",
       "      <td>245</td>\n",
       "      <td>205.0</td>\n",
       "      <td>540.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>fig|426786.38.CDS.2</td>\n",
       "      <td>Small Nonstructural Protein NSs (NSs Protein)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>226</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>fig|426786.38.CDS.3</td>\n",
       "      <td>Pre-glycoprotein polyprotein GP complex (GPC p...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1366</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>fig|426786.38.CDS.4</td>\n",
       "      <td>RNA-dependent RNA polymerase (L protein)</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2083</td>\n",
       "      <td>1719.0</td>\n",
       "      <td>3211.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>fig|426786.38.mat_peptide.1</td>\n",
       "      <td>Phenuiviridae mature nonstructural 78-kD protein</td>\n",
       "      <td>NaN</td>\n",
       "      <td>647</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>fig|426786.38.mat_peptide.2</td>\n",
       "      <td>Mature envelope glycoprotein Gn (Gn protein)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>512</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>fig|426786.38.mat_peptide.3</td>\n",
       "      <td>Mature envelope glycoprotein Gc (Gc protein)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>506</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                            ID  \\\n",
       "0          fig|426786.38.CDS.1   \n",
       "1          fig|426786.38.CDS.2   \n",
       "2          fig|426786.38.CDS.3   \n",
       "3          fig|426786.38.CDS.4   \n",
       "4  fig|426786.38.mat_peptide.1   \n",
       "5  fig|426786.38.mat_peptide.2   \n",
       "6  fig|426786.38.mat_peptide.3   \n",
       "\n",
       "                                            Function  Evaluated  Prot_Len  \\\n",
       "0                   Nucleocapsid protein (N protein)        1.0       245   \n",
       "1      Small Nonstructural Protein NSs (NSs Protein)        NaN       226   \n",
       "2  Pre-glycoprotein polyprotein GP complex (GPC p...        NaN      1366   \n",
       "3           RNA-dependent RNA polymerase (L protein)        1.0      2083   \n",
       "4   Phenuiviridae mature nonstructural 78-kD protein        NaN       647   \n",
       "5       Mature envelope glycoprotein Gn (Gn protein)        NaN       512   \n",
       "6       Mature envelope glycoprotein Gc (Gc protein)        NaN       506   \n",
       "\n",
       "   Min_Len  Max_Len  Copy_Num  Exp_Copy_Num  Exceptions  \n",
       "0    205.0    540.0       1.0           1.0         NaN  \n",
       "1      NaN      NaN       NaN           NaN         NaN  \n",
       "2      NaN      NaN       NaN           NaN         NaN  \n",
       "3   1719.0   3211.0       1.0           1.0         NaN  \n",
       "4      NaN      NaN       NaN           NaN         NaN  \n",
       "5      NaN      NaN       NaN           NaN         NaN  \n",
       "6      NaN      NaN       NaN           NaN         NaN  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# feature_quality\n",
    "print(f\"Total files: {len(sorted(quality_gto_dir.glob('*.feature_quality')))}\")\n",
    "ex_file = 'GCF_031497195.1'\n",
    "fq = pd.read_csv(quality_gto_dir / f'{ex_file}.feature_quality', sep='\\t')\n",
    "fq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# qual.gto"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total files: 1572\n",
      "analysis_events: <class 'list'>\n",
      "close_genomes: <class 'list'>\n",
      "contigs: <class 'list'>\n",
      "domain: <class 'str'>\n",
      "features: <class 'list'>\n",
      "genetic_code: <class 'int'>\n",
      "id: <class 'str'>\n",
      "ncbi_taxonomy_id: <class 'int'>\n",
      "quality: <class 'dict'>\n",
      "scientific_name: <class 'str'>\n"
     ]
    }
   ],
   "source": [
    "# qual.gto\n",
    "print(f\"Total files: {len(sorted(quality_gto_dir.glob('*.qual.gto')))}\")\n",
    "ex_file = 'GCF_031497195.1'\n",
    "file_path = quality_gto_dir / f'{ex_file}.qual.gto'\n",
    "\n",
    "# Load the GTP file\n",
    "with open(file_path, 'r') as file:\n",
    "    gto = json.load(file)\n",
    "\n",
    "for k in sorted(gto.keys()):\n",
    "    print(f'{k}: {type(gto[k])}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7\n",
      "['annotations', 'family_assignments', 'feature_creation_event', 'function', 'id', 'location', 'protein_translation', 'type']\n",
      "['annotations', 'family_assignments', 'feature_creation_event', 'function', 'id', 'location', 'protein_translation', 'type']\n",
      "['annotations', 'family_assignments', 'feature_creation_event', 'function', 'id', 'location', 'protein_translation', 'type']\n",
      "['annotations', 'family_assignments', 'feature_creation_event', 'function', 'id', 'location', 'protein_translation', 'type']\n",
      "['annotations', 'family_assignments', 'feature_creation_event', 'function', 'id', 'location', 'protein_translation', 'type']\n",
      "['annotations', 'family_assignments', 'feature_creation_event', 'function', 'id', 'location', 'protein_translation', 'type']\n",
      "['annotations', 'family_assignments', 'feature_creation_event', 'function', 'id', 'location', 'protein_translation', 'type']\n"
     ]
    }
   ],
   "source": [
    "# Explore 'features' in the GTO dict\n",
    "print(len(gto['features'])) # that's the number rows in feature_quality file\n",
    "for d in gto['features']:\n",
    "    print(sorted(d.keys())) # print sorted keys\n",
    "\n",
    "# create a list of sorted dicts\n",
    "fea = []\n",
    "for d in gto['features']:\n",
    "    fea.append({k: d[k] for k in sorted(d.keys())})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'annotations': [['Add feature',\n",
       "   'annotate_by_viral_pssm',\n",
       "   1742556319.00236,\n",
       "   '2B74E144-0647-11F0-A1B3-509768D45E13'],\n",
       "  ['Set function to Nucleocapsid protein (N protein)',\n",
       "   'annotate_by_viral_pssm',\n",
       "   1742556319.00236,\n",
       "   '2B74E144-0647-11F0-A1B3-509768D45E13']],\n",
       " 'family_assignments': [['Phenuiviridae',\n",
       "   'Phenuiviridae.N.3.pssm',\n",
       "   'Nucleocapsid protein (N protein)',\n",
       "   'annotate_by_viral_pssm']],\n",
       " 'feature_creation_event': '2B74E144-0647-11F0-A1B3-509768D45E13',\n",
       " 'function': 'Nucleocapsid protein (N protein)',\n",
       " 'id': 'fig|426786.38.CDS.1',\n",
       " 'location': [['NC_086346.1', '70', '+', '738']],\n",
       " 'protein_translation': 'MTDYADIAIAFAGEPINNAEVMGWVNEFAYEGFNAQRIIQLVQEKGPQTWQTDVKMMIVLALTRGNKPSKMIEKMSAEGKKKASRLITIYGLKSGNPGRDDLTLSRIAAAFAGWTCQALATLHPYLPVTGAAMDAISPGYPRAMMHPSFAGLIDNSIPEAYLQVVVDAHALYLLQFSRVINRNMRGQPKSVVVSSFLQPMNAAIVSGFISNDRRRKMLMAFGIVDQNGKPTAAVESAAKAFMTAV*',\n",
       " 'type': 'CDS'}"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fea[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "fig|426786.38.CDS.1,  CDS,  Nucleocapsid protein (N protein)\n",
      "fig|426786.38.CDS.2,  CDS,  Small Nonstructural Protein NSs (NSs Protein)\n",
      "fig|426786.38.CDS.3,  CDS,  Pre-glycoprotein polyprotein GP complex (GPC protein)\n",
      "fig|426786.38.CDS.4,  CDS,  RNA-dependent RNA polymerase (L protein)\n",
      "fig|426786.38.mat_peptide.1,  mat_peptide,  Phenuiviridae mature nonstructural 78-kD protein\n",
      "fig|426786.38.mat_peptide.2,  mat_peptide,  Mature envelope glycoprotein Gn (Gn protein)\n",
      "fig|426786.38.mat_peptide.3,  mat_peptide,  Mature envelope glycoprotein Gc (Gc protein)\n"
     ]
    }
   ],
   "source": [
    "for i, item in enumerate(fea):\n",
    "    print(f\"{item['id']},  {item['type']},  {item['function']}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Summary data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "# def summarize_data_from_gto(gto_file_path):\n",
    "#     \"\"\"\n",
    "#     Extract protein data and metadata from a GTO file and return it as a DataFrame.\n",
    "#     \"\"\"\n",
    "#     # file_path = quality_gto_dir / 'GCF_031497195.1.qual.gto'\n",
    "\n",
    "#     ## Load the GTO file\n",
    "#     with open(gto_file_path, 'r') as file:\n",
    "#         gto = json.load(file)\n",
    "\n",
    "#     ## 'features' key in GTO\n",
    "#     # Each item in 'features' is a dict with keys: 'id', 'type', 'function', etc.\n",
    "#     # fea is a list of dicts, each representing a 'features' item\n",
    "#     fea = []\n",
    "#     for d in gto['features']:\n",
    "#         fea.append({k: d[k] for k in sorted(d.keys())})\n",
    "\n",
    "#     # Aggregate the 'features' items into a DataFrame\n",
    "#     df = []\n",
    "#     # features_columns contains available data items from 'features' key\n",
    "#     features_columns = ['id', 'type', 'function', 'protein_translation']\n",
    "#     for i, item in enumerate(fea):\n",
    "#         df.append([item[f] if f in item else np.nan for f in features_columns])\n",
    "#         # df.append([item[f] for f in features_columns])\n",
    "#     df = pd.DataFrame(df, columns=features_columns)\n",
    "#     df['length'] = df['protein_translation'].apply(lambda x: len(x) if isinstance(x, str) else 0)\n",
    "\n",
    "#     ## General, GTO file level metadata\n",
    "#     df.insert(0, 'quality', gto['quality']['genome_quality'], allow_duplicates=True)\n",
    "#     df.insert(0, 'file', gto_file_path.name, allow_duplicates=True)\n",
    "#     return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Protein (feature) data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_protein_data_from_gto(gto_file_path):\n",
    "    \"\"\"\n",
    "    Extract protein data and metadata from a GTO file and return it as a DataFrame.\n",
    "    \"\"\"\n",
    "    # file_path = quality_gto_dir / 'GCF_031497195.1.qual.gto'\n",
    "\n",
    "    ## Load the GTO file\n",
    "    with open(gto_file_path, 'r') as file:\n",
    "        gto = json.load(file)\n",
    "\n",
    "    ## 'features' key in GTO\n",
    "    # Each item in 'features' is a dict with keys: 'id', 'type', 'function', etc.\n",
    "    # fea is a list of dicts, each representing a 'features' item\n",
    "    fea = []\n",
    "    for d in gto['features']:\n",
    "        fea.append({k: d[k] for k in sorted(d.keys())})\n",
    "\n",
    "    # Aggregate the 'features' items into a DataFrame\n",
    "    df = []\n",
    "    # features_columns contains available data items from 'features' key\n",
    "    features_columns = ['id', 'type', 'function', 'protein_translation']\n",
    "    for i, item in enumerate(fea):\n",
    "        df.append([item[f] if f in item else np.nan for f in features_columns])\n",
    "        # df.append([item[f] for f in features_columns])\n",
    "    df = pd.DataFrame(df, columns=features_columns)\n",
    "    df['length'] = df['protein_translation'].apply(lambda x: len(x) if isinstance(x, str) else 0)\n",
    "\n",
    "    ## General, GTO file level metadata\n",
    "    df.insert(0, 'quality', gto['quality']['genome_quality'], allow_duplicates=True)\n",
    "    df.insert(0, 'file', gto_file_path.name, allow_duplicates=True)\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>file</th>\n",
       "      <th>quality</th>\n",
       "      <th>id</th>\n",
       "      <th>type</th>\n",
       "      <th>function</th>\n",
       "      <th>protein_translation</th>\n",
       "      <th>length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GCA_000851025.1.qual.gto</td>\n",
       "      <td>Poor</td>\n",
       "      <td>fig|3052303.150.mat_peptide.1</td>\n",
       "      <td>mat_peptide</td>\n",
       "      <td>Mature envelope glycoprotein Gc (Gc protein)</td>\n",
       "      <td>GTFTWTLSDSSGVENPGGYCLTKWMILAAELKCFGNTAVAKCNVNH...</td>\n",
       "      <td>234</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GCA_000851025.1.qual.gto</td>\n",
       "      <td>Poor</td>\n",
       "      <td>fig|3052303.150.CDS.1</td>\n",
       "      <td>CDS</td>\n",
       "      <td>RNA-dependent RNA polymerase (L protein)</td>\n",
       "      <td>MDEIISELRELCLNYIEQDERLSRQKLNFLGQREPRMVLIEGLKLL...</td>\n",
       "      <td>2211</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GCA_000851025.1.qual.gto</td>\n",
       "      <td>Poor</td>\n",
       "      <td>fig|3052303.150.CDS.2</td>\n",
       "      <td>CDS</td>\n",
       "      <td>Pre-glycoprotein polyprotein GP complex (GPC p...</td>\n",
       "      <td>MGQIVTMFEALPHIIDEVINIVIIVLIVITGIKAVYNFATCGIFAL...</td>\n",
       "      <td>499</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GCA_000851025.1.qual.gto</td>\n",
       "      <td>Poor</td>\n",
       "      <td>fig|3052303.150.CDS.3</td>\n",
       "      <td>CDS</td>\n",
       "      <td>Nucleocapsid protein (N protein)</td>\n",
       "      <td>MSLSKEVKSFQWTQALRRELQSFTSDVKAAVIKDATNLLNGLDFSE...</td>\n",
       "      <td>559</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GCA_000851025.1.qual.gto</td>\n",
       "      <td>Poor</td>\n",
       "      <td>fig|3052303.150.RNA.1</td>\n",
       "      <td>RNA</td>\n",
       "      <td>Small segment stemloop region</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       file quality                             id  \\\n",
       "0  GCA_000851025.1.qual.gto    Poor  fig|3052303.150.mat_peptide.1   \n",
       "1  GCA_000851025.1.qual.gto    Poor          fig|3052303.150.CDS.1   \n",
       "2  GCA_000851025.1.qual.gto    Poor          fig|3052303.150.CDS.2   \n",
       "3  GCA_000851025.1.qual.gto    Poor          fig|3052303.150.CDS.3   \n",
       "4  GCA_000851025.1.qual.gto    Poor          fig|3052303.150.RNA.1   \n",
       "\n",
       "          type                                           function  \\\n",
       "0  mat_peptide       Mature envelope glycoprotein Gc (Gc protein)   \n",
       "1          CDS           RNA-dependent RNA polymerase (L protein)   \n",
       "2          CDS  Pre-glycoprotein polyprotein GP complex (GPC p...   \n",
       "3          CDS                   Nucleocapsid protein (N protein)   \n",
       "4          RNA                      Small segment stemloop region   \n",
       "\n",
       "                                 protein_translation  length  \n",
       "0  GTFTWTLSDSSGVENPGGYCLTKWMILAAELKCFGNTAVAKCNVNH...     234  \n",
       "1  MDEIISELRELCLNYIEQDERLSRQKLNFLGQREPRMVLIEGLKLL...    2211  \n",
       "2  MGQIVTMFEALPHIIDEVINIVIIVLIVITGIKAVYNFATCGIFAL...     499  \n",
       "3  MSLSKEVKSFQWTQALRRELQSFTSDVKAAVIKDATNLLNGLDFSE...     559  \n",
       "4                                                NaN       0  "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# ex_file = 'GCF_031497195.1'\n",
    "ex_file = 'GCA_000851025.1'\n",
    "df = get_protein_data_from_gto(gto_file_path=quality_gto_dir / f'{ex_file}.qual.gto')\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total files: 1572\n",
      "Creating /nfs/lambda_stor_01/data/apartin/projects/cepi/viral-segmatch/notebooks/../data/bunya_processed/protein_data.tsv\n",
      "(8400, 7)\n",
      "['file', 'quality', 'id', 'type', 'function', 'protein_translation', 'length']\n"
     ]
    }
   ],
   "source": [
    "# Aggregate protein data from all GTO files\n",
    "print(f\"Total files: {len(sorted(quality_gto_dir.glob('*.qual.gto')))}\")\n",
    "\n",
    "protein_data_fname = 'protein_data.tsv'\n",
    "\n",
    "if not (output_dir / protein_data_fname).exists():\n",
    "    print(f'Creating {output_dir / protein_data_fname}')\n",
    "    dfs = []    \n",
    "    for i, fpath in enumerate(sorted(quality_gto_dir.glob('*.gto'))):\n",
    "        df = get_protein_data_from_gto(gto_file_path=fpath)\n",
    "        dfs.append(df)\n",
    "\n",
    "    prot_df = pd.concat(dfs, axis=0).reset_index(drop=True)\n",
    "    prot_df.to_csv(output_dir / protein_data_fname, sep='\\t', index=False)\n",
    "\n",
    "else:\n",
    "    print(f'Loading {output_dir / protein_data_fname}')\n",
    "    prot_df = pd.read_csv(output_dir / protein_data_fname, sep='\\t')\n",
    "\n",
    "print(prot_df.shape)\n",
    "print(prot_df.columns.tolist())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(184, 7)\n"
     ]
    }
   ],
   "source": [
    "# Save DataFrame with samples that don't have protein data\n",
    "no_prot_df = prot_df[prot_df['protein_translation'].isna()]\n",
    "print(no_prot_df.shape)\n",
    "no_prot_df.to_csv(output_dir / 'missing_protein_data.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(6646, 7)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>file</th>\n",
       "      <th>quality</th>\n",
       "      <th>id</th>\n",
       "      <th>type</th>\n",
       "      <th>function</th>\n",
       "      <th>protein_translation</th>\n",
       "      <th>length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5890</th>\n",
       "      <td>GCF_000871205.1.qual.gto</td>\n",
       "      <td>Good</td>\n",
       "      <td>fig|70566.1857.mat_peptide.3</td>\n",
       "      <td>mat_peptide</td>\n",
       "      <td>Mature envelope glycoprotein Gc (Gc protein)</td>\n",
       "      <td>ACIQEKEIETIDDAAMCIGLYQNITQPKEYNTFVKELSSTLSSHEI...</td>\n",
       "      <td>936</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>250</th>\n",
       "      <td>GCA_000871205.1.qual.gto</td>\n",
       "      <td>Good</td>\n",
       "      <td>fig|70566.1858.mat_peptide.3</td>\n",
       "      <td>mat_peptide</td>\n",
       "      <td>Mature envelope glycoprotein Gc (Gc protein)</td>\n",
       "      <td>ACIQEKEIETIDDAAMCIGLYQNITQPKEYNTFVKELSSTLSSHEI...</td>\n",
       "      <td>936</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6838</th>\n",
       "      <td>GCF_002925575.1.qual.gto</td>\n",
       "      <td>Poor</td>\n",
       "      <td>fig|1145238.48.mat_peptide.2</td>\n",
       "      <td>mat_peptide</td>\n",
       "      <td>Mature envelope glycoprotein Gc (Gc protein)</td>\n",
       "      <td>ADTPILEPGWSDTAHGVGEIPMKTDLELDFSLPSSSSYSYRRKLTN...</td>\n",
       "      <td>478</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1220</th>\n",
       "      <td>GCA_002925575.1.qual.gto</td>\n",
       "      <td>Poor</td>\n",
       "      <td>fig|1145238.49.mat_peptide.2</td>\n",
       "      <td>mat_peptide</td>\n",
       "      <td>Mature envelope glycoprotein Gc (Gc protein)</td>\n",
       "      <td>ADTPILEPGWSDTAHGVGEIPMKTDLELDFSLPSSSSYSYRRKLTN...</td>\n",
       "      <td>478</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>839</th>\n",
       "      <td>GCA_002117715.1.qual.gto</td>\n",
       "      <td>Good</td>\n",
       "      <td>fig|3052486.14.mat_peptide.2</td>\n",
       "      <td>mat_peptide</td>\n",
       "      <td>Mature envelope glycoprotein Gc (Gc protein)</td>\n",
       "      <td>ADTPLAEVGWSDTAHGVGVIPLKTDLELDFALASSASYSYRRKLSN...</td>\n",
       "      <td>476</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6475</th>\n",
       "      <td>GCF_002117715.1.qual.gto</td>\n",
       "      <td>Good</td>\n",
       "      <td>fig|3052486.13.mat_peptide.2</td>\n",
       "      <td>mat_peptide</td>\n",
       "      <td>Mature envelope glycoprotein Gc (Gc protein)</td>\n",
       "      <td>ADTPLAEVGWSDTAHGVGVIPLKTDLELDFALASSASYSYRRKLSN...</td>\n",
       "      <td>476</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6683</th>\n",
       "      <td>GCF_002817355.1.qual.gto</td>\n",
       "      <td>Poor</td>\n",
       "      <td>fig|3052490.13.mat_peptide.2</td>\n",
       "      <td>mat_peptide</td>\n",
       "      <td>Mature envelope glycoprotein Gc (Gc protein)</td>\n",
       "      <td>ADTPLLEPGWSDTAHGVGDIPMKTDLELDFAIPSSSSYSYRRRLVN...</td>\n",
       "      <td>476</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1050</th>\n",
       "      <td>GCA_002817355.1.qual.gto</td>\n",
       "      <td>Poor</td>\n",
       "      <td>fig|3052490.14.mat_peptide.2</td>\n",
       "      <td>mat_peptide</td>\n",
       "      <td>Mature envelope glycoprotein Gc (Gc protein)</td>\n",
       "      <td>ADTPLLEPGWSDTAHGVGDIPMKTDLELDFAIPSSSSYSYRRRLVN...</td>\n",
       "      <td>476</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6740</th>\n",
       "      <td>GCF_002830985.1.qual.gto</td>\n",
       "      <td>Good</td>\n",
       "      <td>fig|3052499.526.mat_peptide.2</td>\n",
       "      <td>mat_peptide</td>\n",
       "      <td>Mature envelope glycoprotein Gc (Gc protein)</td>\n",
       "      <td>ADTPLMESGWSDTAHGVGIIPMKTDLELDFALASSSSYSYRRKLVN...</td>\n",
       "      <td>476</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>133</th>\n",
       "      <td>GCA_000854765.2.qual.gto</td>\n",
       "      <td>Poor</td>\n",
       "      <td>fig|3052499.527.mat_peptide.2</td>\n",
       "      <td>mat_peptide</td>\n",
       "      <td>Mature envelope glycoprotein Gc (Gc protein)</td>\n",
       "      <td>ADTPLMESGWSDTAHGVGIIPMKTDLELDFALASSSSYSYRRKLVN...</td>\n",
       "      <td>476</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1112</th>\n",
       "      <td>GCA_002830985.1.qual.gto</td>\n",
       "      <td>Good</td>\n",
       "      <td>fig|3052499.529.mat_peptide.2</td>\n",
       "      <td>mat_peptide</td>\n",
       "      <td>Mature envelope glycoprotein Gc (Gc protein)</td>\n",
       "      <td>ADTPLMESGWSDTAHGVGIIPMKTDLELDFALASSSSYSYRRKLVN...</td>\n",
       "      <td>476</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5769</th>\n",
       "      <td>GCF_000854765.1.qual.gto</td>\n",
       "      <td>Good</td>\n",
       "      <td>fig|3052499.525.mat_peptide.2</td>\n",
       "      <td>mat_peptide</td>\n",
       "      <td>Mature envelope glycoprotein Gc (Gc protein)</td>\n",
       "      <td>ADTPLMESGWSDTAHGVGIIPMKTDLELDFALASSSSYSYRRKLVN...</td>\n",
       "      <td>476</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                          file quality                             id  \\\n",
       "5890  GCF_000871205.1.qual.gto    Good   fig|70566.1857.mat_peptide.3   \n",
       "250   GCA_000871205.1.qual.gto    Good   fig|70566.1858.mat_peptide.3   \n",
       "6838  GCF_002925575.1.qual.gto    Poor   fig|1145238.48.mat_peptide.2   \n",
       "1220  GCA_002925575.1.qual.gto    Poor   fig|1145238.49.mat_peptide.2   \n",
       "839   GCA_002117715.1.qual.gto    Good   fig|3052486.14.mat_peptide.2   \n",
       "6475  GCF_002117715.1.qual.gto    Good   fig|3052486.13.mat_peptide.2   \n",
       "6683  GCF_002817355.1.qual.gto    Poor   fig|3052490.13.mat_peptide.2   \n",
       "1050  GCA_002817355.1.qual.gto    Poor   fig|3052490.14.mat_peptide.2   \n",
       "6740  GCF_002830985.1.qual.gto    Good  fig|3052499.526.mat_peptide.2   \n",
       "133   GCA_000854765.2.qual.gto    Poor  fig|3052499.527.mat_peptide.2   \n",
       "1112  GCA_002830985.1.qual.gto    Good  fig|3052499.529.mat_peptide.2   \n",
       "5769  GCF_000854765.1.qual.gto    Good  fig|3052499.525.mat_peptide.2   \n",
       "\n",
       "             type                                      function  \\\n",
       "5890  mat_peptide  Mature envelope glycoprotein Gc (Gc protein)   \n",
       "250   mat_peptide  Mature envelope glycoprotein Gc (Gc protein)   \n",
       "6838  mat_peptide  Mature envelope glycoprotein Gc (Gc protein)   \n",
       "1220  mat_peptide  Mature envelope glycoprotein Gc (Gc protein)   \n",
       "839   mat_peptide  Mature envelope glycoprotein Gc (Gc protein)   \n",
       "6475  mat_peptide  Mature envelope glycoprotein Gc (Gc protein)   \n",
       "6683  mat_peptide  Mature envelope glycoprotein Gc (Gc protein)   \n",
       "1050  mat_peptide  Mature envelope glycoprotein Gc (Gc protein)   \n",
       "6740  mat_peptide  Mature envelope glycoprotein Gc (Gc protein)   \n",
       "133   mat_peptide  Mature envelope glycoprotein Gc (Gc protein)   \n",
       "1112  mat_peptide  Mature envelope glycoprotein Gc (Gc protein)   \n",
       "5769  mat_peptide  Mature envelope glycoprotein Gc (Gc protein)   \n",
       "\n",
       "                                    protein_translation  length  \n",
       "5890  ACIQEKEIETIDDAAMCIGLYQNITQPKEYNTFVKELSSTLSSHEI...     936  \n",
       "250   ACIQEKEIETIDDAAMCIGLYQNITQPKEYNTFVKELSSTLSSHEI...     936  \n",
       "6838  ADTPILEPGWSDTAHGVGEIPMKTDLELDFSLPSSSSYSYRRKLTN...     478  \n",
       "1220  ADTPILEPGWSDTAHGVGEIPMKTDLELDFSLPSSSSYSYRRKLTN...     478  \n",
       "839   ADTPLAEVGWSDTAHGVGVIPLKTDLELDFALASSASYSYRRKLSN...     476  \n",
       "6475  ADTPLAEVGWSDTAHGVGVIPLKTDLELDFALASSASYSYRRKLSN...     476  \n",
       "6683  ADTPLLEPGWSDTAHGVGDIPMKTDLELDFAIPSSSSYSYRRRLVN...     476  \n",
       "1050  ADTPLLEPGWSDTAHGVGDIPMKTDLELDFAIPSSSSYSYRRRLVN...     476  \n",
       "6740  ADTPLMESGWSDTAHGVGIIPMKTDLELDFALASSSSYSYRRKLVN...     476  \n",
       "133   ADTPLMESGWSDTAHGVGIIPMKTDLELDFALASSSSYSYRRKLVN...     476  \n",
       "1112  ADTPLMESGWSDTAHGVGIIPMKTDLELDFALASSSSYSYRRKLVN...     476  \n",
       "5769  ADTPLMESGWSDTAHGVGIIPMKTDLELDFALASSSSYSYRRKLVN...     476  "
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check duplicates\n",
    "dups = prot_df[prot_df.duplicated(subset=['protein_translation'], keep=False)].sort_values(['protein_translation'])\n",
    "dups.to_csv(output_dir / 'protein_duplicates.tsv', sep='\\t', index=False)\n",
    "print(dups.shape)\n",
    "dups[:12]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# DNA/RNA (contigs) data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_dna_data_from_gto(gto_file_path):\n",
    "    \"\"\"\n",
    "    Extract dna/rna data and metadata from a GTO file and return it as a DataFrame.\n",
    "    \"\"\"\n",
    "    # file_path = quality_gto_dir / 'GCF_031497195.1.qual.gto'\n",
    "\n",
    "    ## Load the GTO file\n",
    "    with open(gto_file_path, 'r') as file:\n",
    "        gto = json.load(file)\n",
    "\n",
    "    ## 'contigs' key in GTO\n",
    "    # Each item in 'contigs' is a dict with keys: 'id', 'replicon_type', 'replicon_geometry', 'dna'\n",
    "    # ctg is a list of dicts, each representing a 'contigs' item\n",
    "    cng = []\n",
    "    for d in gto['contigs']:\n",
    "        cng.append({k: d[k] for k in sorted(d.keys())})\n",
    "\n",
    "    # Aggregate the 'contigs' items into a DataFrame\n",
    "    df = []\n",
    "    # contigs_columns contains available data items from 'contigs' key\n",
    "    contigs_columns = ['id', 'replicon_type', 'replicon_geometry', 'dna']\n",
    "    for i, item in enumerate(cng):\n",
    "        df.append([item[f] if f in item else np.nan for f in contigs_columns])\n",
    "    df = pd.DataFrame(df, columns=contigs_columns)\n",
    "    df['length'] = df['dna'].apply(lambda x: len(x) if isinstance(x, str) else 0)\n",
    "\n",
    "    ## General, GTO file level metadata\n",
    "    df.insert(0, 'quality', gto['quality']['genome_quality'], allow_duplicates=True)\n",
    "    df.insert(0, 'file', gto_file_path.name, allow_duplicates=True)\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total files: 1572\n",
      "Creating /nfs/lambda_stor_01/data/apartin/projects/cepi/viral-segmatch/notebooks/../data/bunya_processed/dna_data.tsv\n",
      "(4686, 7)\n",
      "['file', 'quality', 'id', 'replicon_type', 'replicon_geometry', 'dna', 'length']\n"
     ]
    }
   ],
   "source": [
    "# Aggregate dna data from all GTO files\n",
    "print(f\"Total files: {len(sorted(quality_gto_dir.glob('*.qual.gto')))}\")\n",
    "\n",
    "dna_data_fname = 'dna_data.tsv'\n",
    "\n",
    "if not (output_dir / dna_data_fname).exists():\n",
    "    print(f'Creating {output_dir / dna_data_fname}')\n",
    "    dfs = []    \n",
    "    for i, fpath in enumerate(sorted(quality_gto_dir.glob('*.gto'))):\n",
    "        df = get_dna_data_from_gto(gto_file_path=fpath)\n",
    "        dfs.append(df)\n",
    "\n",
    "    dna_df = pd.concat(dfs, axis=0).reset_index(drop=True)\n",
    "    dna_df.to_csv(output_dir / dna_data_fname, sep='\\t', index=False)\n",
    "\n",
    "else:\n",
    "    print(f'Loading {output_dir / dna_data_fname}')\n",
    "    dna_df = pd.read_csv(output_dir / dna_data_fname, sep='\\t')\n",
    "\n",
    "print(dna_df.shape)\n",
    "print(dna_df.columns.tolist())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(0, 7)\n"
     ]
    }
   ],
   "source": [
    "# Save DataFrame with samples that don't have protein data\n",
    "no_dna_df = dna_df[dna_df['dna'].isna()]\n",
    "print(no_dna_df.shape)\n",
    "no_dna_df.to_csv(output_dir / 'missing_dna_data.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "# df = dna_df.drop(columns=['dna', 'length'])\n",
    "# df.to_csv(output_dir / 'dna_data_no_seq.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1. Check quality control in DNA data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>file</th>\n",
       "      <th>quality</th>\n",
       "      <th>id</th>\n",
       "      <th>replicon_type</th>\n",
       "      <th>replicon_geometry</th>\n",
       "      <th>dna</th>\n",
       "      <th>length</th>\n",
       "      <th>ambiguous_count</th>\n",
       "      <th>ambiguous_fraction</th>\n",
       "      <th>length</th>\n",
       "      <th>gc_content</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GCA_000847345.1.qual.gto</td>\n",
       "      <td>Good</td>\n",
       "      <td>DQ375403.1</td>\n",
       "      <td>Large RNA Segment</td>\n",
       "      <td>linear</td>\n",
       "      <td>ACACAAAGGCGCCCAATCATGGATTCTATATTATCAAAACAGCTGG...</td>\n",
       "      <td>6404</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6404</td>\n",
       "      <td>0.438944</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GCA_000847345.1.qual.gto</td>\n",
       "      <td>Good</td>\n",
       "      <td>DQ380206.1</td>\n",
       "      <td>Medium RNA Segment</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ACACAAAGACGGTGCATTAAATGTATGTTTTATTAACAATTCTAAT...</td>\n",
       "      <td>3885</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3885</td>\n",
       "      <td>0.454569</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GCA_000847345.1.qual.gto</td>\n",
       "      <td>Good</td>\n",
       "      <td>DQ380151.1</td>\n",
       "      <td>Small RNA Segment</td>\n",
       "      <td>linear</td>\n",
       "      <td>ACACAAAGACCCCCTAGTGCTTATCAAGTATATCATGGATTACTTT...</td>\n",
       "      <td>1690</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1690</td>\n",
       "      <td>0.489349</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       file quality          id       replicon_type  \\\n",
       "0  GCA_000847345.1.qual.gto    Good  DQ375403.1   Large RNA Segment   \n",
       "1  GCA_000847345.1.qual.gto    Good  DQ380206.1  Medium RNA Segment   \n",
       "2  GCA_000847345.1.qual.gto    Good  DQ380151.1   Small RNA Segment   \n",
       "\n",
       "  replicon_geometry                                                dna  \\\n",
       "0            linear  ACACAAAGGCGCCCAATCATGGATTCTATATTATCAAAACAGCTGG...   \n",
       "1               NaN  ACACAAAGACGGTGCATTAAATGTATGTTTTATTAACAATTCTAAT...   \n",
       "2            linear  ACACAAAGACCCCCTAGTGCTTATCAAGTATATCATGGATTACTTT...   \n",
       "\n",
       "   length  ambiguous_count  ambiguous_fraction  length  gc_content  \n",
       "0    6404                0                 0.0    6404    0.438944  \n",
       "1    3885                0                 0.0    3885    0.454569  \n",
       "2    1690                0                 0.0    1690    0.489349  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ambiguous_fraction\n",
      "0.169765       1\n",
      "0.033234       2\n",
      "0.028332       1\n",
      "0.023320       1\n",
      "0.019280       1\n",
      "            ... \n",
      "0.000138       2\n",
      "0.000113       1\n",
      "0.000083       1\n",
      "0.000082       2\n",
      "0.000000    4457\n",
      "Name: count, Length: 147, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "# DNA Quality Control\n",
    "def summarize_dna_qc(df, seq_col='dna'):\n",
    "    \"\"\" Perform quality control on DNA sequences in a DataFrame. \"\"\"\n",
    "    df = df.copy()\n",
    "    ambiguous_codes = set(\"NRYWSKMBDHV\")  # IUPAC ambiguous nucleotide codes\n",
    "    \n",
    "    qc = []\n",
    "    for idx, seq in df[seq_col].items():\n",
    "        if not isinstance(seq, str) or len(seq) == 0:\n",
    "            qc.append((0, 0.0, 0, 0.0))  # Handle empty or missing sequences\n",
    "            continue\n",
    "        \n",
    "        total_length = len(seq)\n",
    "        ambiguous_count = sum(1 for base in seq if base in ambiguous_codes)\n",
    "        ambiguous_fraction = ambiguous_count / total_length\n",
    "        gc_content = (seq.count('G') + seq.count('C')) / total_length\n",
    "        \n",
    "        qc.append((ambiguous_count, ambiguous_fraction, total_length, gc_content))\n",
    "    \n",
    "    columns = ['ambiguous_count', 'ambiguous_fraction', 'length', 'gc_content']\n",
    "    qc_df = pd.DataFrame(qc, columns=columns).reset_index(drop=True)\n",
    "    df_combined = pd.concat([df, qc_df], axis=1)\n",
    "    # return df, qc_df, df_combined\n",
    "    return df_combined\n",
    "\n",
    "dna_qc = summarize_dna_qc(dna_df, seq_col='dna')\n",
    "\n",
    "# There is only sample with ambiguous_fraction of 0.169765. The rest are less than 0.035.\n",
    "# TODO! How should we filter based on QC??\n",
    "display(dna_qc[:3])\n",
    "print(dna_qc['ambiguous_fraction'].value_counts().sort_index(ascending=False))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2. Check naming consistency in DNA data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total ids:  4686\n",
      "Unique ids: 4686\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "replicon_type\n",
       "Large RNA Segment     1503\n",
       "Small RNA Segment     1371\n",
       "Medium RNA Segment    1283\n",
       "NaN                    311\n",
       "Segment One             61\n",
       "Segment Two             59\n",
       "Segment Four            53\n",
       "Segment Three           45\n",
       "Name: count, dtype: int64"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# All segment IDs are unique (no duplicates).\n",
    "print(f\"Total ids:  {dna_df['id'].count()}\")\n",
    "print(f\"Unique ids: {dna_df['id'].nunique()}\")\n",
    "\n",
    "# Most segments follow the Large/Medium/Small RNA Segment convention.\n",
    "# However, some segments are labeled as \"Segment One\", \"Segment Two\", ...\n",
    "# 311 entries have missing values\n",
    "# TODO! 1) How can we standardize the \"Segment ...\" labels? 2) How can we handle the missing values?\n",
    "display(dna_df['replicon_type'].value_counts(dropna=False))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3. Check for duplicates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(3208, 9)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>file</th>\n",
       "      <th>quality</th>\n",
       "      <th>id</th>\n",
       "      <th>replicon_type</th>\n",
       "      <th>replicon_geometry</th>\n",
       "      <th>dna</th>\n",
       "      <th>length</th>\n",
       "      <th>assembly_prefix</th>\n",
       "      <th>virus_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2172</th>\n",
       "      <td>GCA_018595275.1.qual.gto</td>\n",
       "      <td>Poor</td>\n",
       "      <td>MW039262.1</td>\n",
       "      <td>Large RNA Segment</td>\n",
       "      <td>linear</td>\n",
       "      <td>AAAAAAAAAAAAAAAAAAAAAAAAAAAGTAGTAAACACACTCATAT...</td>\n",
       "      <td>6855</td>\n",
       "      <td>GCA</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4391</th>\n",
       "      <td>GCF_018595275.1.qual.gto</td>\n",
       "      <td>Poor</td>\n",
       "      <td>NC_078362.1</td>\n",
       "      <td>Large RNA Segment</td>\n",
       "      <td>linear</td>\n",
       "      <td>AAAAAAAAAAAAAAAAAAAAAAAAAAAGTAGTAAACACACTCATAT...</td>\n",
       "      <td>6855</td>\n",
       "      <td>GCF</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2467</th>\n",
       "      <td>GCA_029888265.1.qual.gto</td>\n",
       "      <td>Good</td>\n",
       "      <td>OL774846.1</td>\n",
       "      <td>Large RNA Segment</td>\n",
       "      <td>linear</td>\n",
       "      <td>AAAACAAAAAGAGATGGATCAAGACGTAGATCAAATTCTAAATAGA...</td>\n",
       "      <td>7004</td>\n",
       "      <td>GCA</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4647</th>\n",
       "      <td>GCF_029888265.1.qual.gto</td>\n",
       "      <td>Good</td>\n",
       "      <td>NC_079031.1</td>\n",
       "      <td>Large RNA Segment</td>\n",
       "      <td>linear</td>\n",
       "      <td>AAAACAAAAAGAGATGGATCAAGACGTAGATCAAATTCTAAATAGA...</td>\n",
       "      <td>7004</td>\n",
       "      <td>GCF</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3710</th>\n",
       "      <td>GCF_002831345.1.qual.gto</td>\n",
       "      <td>Good</td>\n",
       "      <td>NC_038733.1</td>\n",
       "      <td>Medium RNA Segment</td>\n",
       "      <td>linear</td>\n",
       "      <td>AAAACAGTAGTGTACCGCTGTGTATAAATAACCTAAGTGATAAAAG...</td>\n",
       "      <td>4411</td>\n",
       "      <td>GCF</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>602</th>\n",
       "      <td>GCA_002831345.1.qual.gto</td>\n",
       "      <td>Good</td>\n",
       "      <td>KP792662.1</td>\n",
       "      <td>Medium RNA Segment</td>\n",
       "      <td>linear</td>\n",
       "      <td>AAAACAGTAGTGTACCGCTGTGTATAAATAACCTAAGTGATAAAAG...</td>\n",
       "      <td>4411</td>\n",
       "      <td>GCA</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4524</th>\n",
       "      <td>GCF_029887365.1.qual.gto</td>\n",
       "      <td>Poor</td>\n",
       "      <td>NC_078877.1</td>\n",
       "      <td>Small RNA Segment</td>\n",
       "      <td>linear</td>\n",
       "      <td>AAAACTTAAAGTCCTTTAATAGTATGGCTGAAGCGATTGTTTTCAA...</td>\n",
       "      <td>910</td>\n",
       "      <td>GCF</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2332</th>\n",
       "      <td>GCA_029887365.1.qual.gto</td>\n",
       "      <td>Poor</td>\n",
       "      <td>MH507156.1</td>\n",
       "      <td>Small RNA Segment</td>\n",
       "      <td>linear</td>\n",
       "      <td>AAAACTTAAAGTCCTTTAATAGTATGGCTGAAGCGATTGTTTTCAA...</td>\n",
       "      <td>910</td>\n",
       "      <td>GCA</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4314</th>\n",
       "      <td>GCF_018594865.1.qual.gto</td>\n",
       "      <td>Good</td>\n",
       "      <td>NC_078254.1</td>\n",
       "      <td>Large RNA Segment</td>\n",
       "      <td>linear</td>\n",
       "      <td>AAAACTTTTTCTCAAAGTTATAAATCCCTTACTCCCCCCTAAACCT...</td>\n",
       "      <td>12189</td>\n",
       "      <td>GCF</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2095</th>\n",
       "      <td>GCA_018594865.1.qual.gto</td>\n",
       "      <td>Good</td>\n",
       "      <td>KU343163.1</td>\n",
       "      <td>Large RNA Segment</td>\n",
       "      <td>linear</td>\n",
       "      <td>AAAACTTTTTCTCAAAGTTATAAATCCCTTACTCCCCCCTAAACCT...</td>\n",
       "      <td>12189</td>\n",
       "      <td>GCA</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4648</th>\n",
       "      <td>GCF_029888265.1.qual.gto</td>\n",
       "      <td>Good</td>\n",
       "      <td>NC_079029.1</td>\n",
       "      <td>Medium RNA Segment</td>\n",
       "      <td>linear</td>\n",
       "      <td>AAAAGTGACTTTCCTGACTTTAACTAACTTTTTGACGAGAGCAATG...</td>\n",
       "      <td>4547</td>\n",
       "      <td>GCF</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2468</th>\n",
       "      <td>GCA_029888265.1.qual.gto</td>\n",
       "      <td>Good</td>\n",
       "      <td>OL774847.1</td>\n",
       "      <td>Medium RNA Segment</td>\n",
       "      <td>linear</td>\n",
       "      <td>AAAAGTGACTTTCCTGACTTTAACTAACTTTTTGACGAGAGCAATG...</td>\n",
       "      <td>4547</td>\n",
       "      <td>GCA</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                          file quality           id       replicon_type  \\\n",
       "2172  GCA_018595275.1.qual.gto    Poor   MW039262.1   Large RNA Segment   \n",
       "4391  GCF_018595275.1.qual.gto    Poor  NC_078362.1   Large RNA Segment   \n",
       "2467  GCA_029888265.1.qual.gto    Good   OL774846.1   Large RNA Segment   \n",
       "4647  GCF_029888265.1.qual.gto    Good  NC_079031.1   Large RNA Segment   \n",
       "3710  GCF_002831345.1.qual.gto    Good  NC_038733.1  Medium RNA Segment   \n",
       "602   GCA_002831345.1.qual.gto    Good   KP792662.1  Medium RNA Segment   \n",
       "4524  GCF_029887365.1.qual.gto    Poor  NC_078877.1   Small RNA Segment   \n",
       "2332  GCA_029887365.1.qual.gto    Poor   MH507156.1   Small RNA Segment   \n",
       "4314  GCF_018594865.1.qual.gto    Good  NC_078254.1   Large RNA Segment   \n",
       "2095  GCA_018594865.1.qual.gto    Good   KU343163.1   Large RNA Segment   \n",
       "4648  GCF_029888265.1.qual.gto    Good  NC_079029.1  Medium RNA Segment   \n",
       "2468  GCA_029888265.1.qual.gto    Good   OL774847.1  Medium RNA Segment   \n",
       "\n",
       "     replicon_geometry                                                dna  \\\n",
       "2172            linear  AAAAAAAAAAAAAAAAAAAAAAAAAAAGTAGTAAACACACTCATAT...   \n",
       "4391            linear  AAAAAAAAAAAAAAAAAAAAAAAAAAAGTAGTAAACACACTCATAT...   \n",
       "2467            linear  AAAACAAAAAGAGATGGATCAAGACGTAGATCAAATTCTAAATAGA...   \n",
       "4647            linear  AAAACAAAAAGAGATGGATCAAGACGTAGATCAAATTCTAAATAGA...   \n",
       "3710            linear  AAAACAGTAGTGTACCGCTGTGTATAAATAACCTAAGTGATAAAAG...   \n",
       "602             linear  AAAACAGTAGTGTACCGCTGTGTATAAATAACCTAAGTGATAAAAG...   \n",
       "4524            linear  AAAACTTAAAGTCCTTTAATAGTATGGCTGAAGCGATTGTTTTCAA...   \n",
       "2332            linear  AAAACTTAAAGTCCTTTAATAGTATGGCTGAAGCGATTGTTTTCAA...   \n",
       "4314            linear  AAAACTTTTTCTCAAAGTTATAAATCCCTTACTCCCCCCTAAACCT...   \n",
       "2095            linear  AAAACTTTTTCTCAAAGTTATAAATCCCTTACTCCCCCCTAAACCT...   \n",
       "4648            linear  AAAAGTGACTTTCCTGACTTTAACTAACTTTTTGACGAGAGCAATG...   \n",
       "2468            linear  AAAAGTGACTTTCCTGACTTTAACTAACTTTTTGACGAGAGCAATG...   \n",
       "\n",
       "      length assembly_prefix virus_id  \n",
       "2172    6855             GCA      NaN  \n",
       "4391    6855             GCF      NaN  \n",
       "2467    7004             GCA      NaN  \n",
       "4647    7004             GCF      NaN  \n",
       "3710    4411             GCF      NaN  \n",
       "602     4411             GCA      NaN  \n",
       "4524     910             GCF      NaN  \n",
       "2332     910             GCA      NaN  \n",
       "4314   12189             GCF      NaN  \n",
       "2095   12189             GCA      NaN  \n",
       "4648    4547             GCF      NaN  \n",
       "2468    4547             GCA      NaN  "
      ]
     },
     "execution_count": 118,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check duplicates\n",
    "# There are duplicates.\n",
    "# TODO! How should we handle the duplicates?\n",
    "# 1. Are these dups necessarily a problem??\n",
    "dna_dups = dna_df[dna_df.duplicated(subset=['dna'], keep=False)].sort_values(['dna'])\n",
    "dna_dups.to_csv(output_dir / 'dna_duplicates.tsv', sep='\\t', index=False)\n",
    "print(dna_dups.shape)\n",
    "dna_dups[:12]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>file</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2</td>\n",
       "      <td>1592</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>5</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   file  count\n",
       "0     2   1592\n",
       "1     3      3\n",
       "2     5      3"
      ]
     },
     "execution_count": 127,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"\"\"\n",
    "1. Count how often each duplicated sequence appears across different files\n",
    "This helps understand if duplicates are common within the same virus or across different ones.\n",
    "\n",
    "Case 1:  2 duplicates.\n",
    "This might be technical (e.g., GCA vs. GCF versions of the same assembly for a given virus). Some GTO files may contain reference (GCF) and draft (GCA) assemblies of the same virus, leading to identical sequences in different files.\n",
    "- If these duplicates exist only between GCA/GCF pairs, we can prefer the GCF version (since it is typically a higher-quality reference).\n",
    "- If they are between two different GCA assemblies, we need to check if they belong to the same virus (see next step).\n",
    "\n",
    "Case 2:  > 2 duplicates.\n",
    "Potential reasons:\n",
    "- Shared sequences across different virus strains (real biological conservation).\n",
    "- Multiple sequencing projects capturing the same virus but submitted separately.\n",
    "- Annotation errors where different assemblies accidentally include redundant sequences.\n",
    "\n",
    "What should we do?\n",
    "- Check if they belong to different viruses (extract and compare assembly IDs, see next section).\n",
    "- If they belong to different viruses, they might be real conserved elements  keep them.\n",
    "- If they belong to the same virus, keep only one high-quality version to avoid redundancy.\n",
    "\n",
    "ap:\n",
    "Most duplicates appear in only 2 files. What does it mean? And, what should we do with those?\n",
    "Then, we have 3 sequences that appear in 3 files, and 3 sequences that appear in 5 files. What does it mean? And, what should we do with those?\n",
    "num_unique_files\tcount\n",
    "2\t1592\n",
    "3\t3\n",
    "5\t3\n",
    "\"\"\"\n",
    "\n",
    "dna_dups = dna_df[dna_df.duplicated(subset=['dna'], keep=False)].sort_values(['dna'])\n",
    "# dna_dups.groupby('dna').agg(file=('file', 'unique')).reset_index().explode('file')\n",
    "\n",
    "dna_dups_count_files = dna_dups[['dna', 'file']].groupby('dna')['file'].nunique().reset_index()\n",
    "dna_dups_count_files['file'].value_counts().reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Follow up on step 1\n",
    "# TODO\n",
    "pass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>assembly_prefix</th>\n",
       "      <th>dna</th>\n",
       "      <th>GCA</th>\n",
       "      <th>GCF</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAAAAAAAAAAAAAAAAAAAAAAAAAAGTAGTAAACACACTCATAT...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAAACAAAAAGAGATGGATCAAGACGTAGATCAAATTCTAAATAGA...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AAAACAGTAGTGTACCGCTGTGTATAAATAACCTAAGTGATAAAAG...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AAAACTTAAAGTCCTTTAATAGTATGGCTGAAGCGATTGTTTTCAA...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AAAACTTTTTCTCAAAGTTATAAATCCCTTACTCCCCCCTAAACCT...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1556</th>\n",
       "      <td>TTTCCATACTACGTGTACGAAGTTTTATATAATAAATAATAAATTA...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1557</th>\n",
       "      <td>TTTCTGAAGAATCGATAAGACAAATCAAAACAAATTAACAATCTCA...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1558</th>\n",
       "      <td>TTTGAATACAACACACATAGACGGCTCTAATTTGAAAAACATAAAA...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1559</th>\n",
       "      <td>TTTGCACCGGGGATCCTAGGCAAAGTGGTTATCTATTTTGGGACTC...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1560</th>\n",
       "      <td>TTTGCTTCATTATACTTGAGCAGTGTGTGACATCTGAGGCAATGTT...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1561 rows  3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "assembly_prefix                                                dna  GCA  GCF\n",
       "0                AAAAAAAAAAAAAAAAAAAAAAAAAAAGTAGTAAACACACTCATAT...  1.0  1.0\n",
       "1                AAAACAAAAAGAGATGGATCAAGACGTAGATCAAATTCTAAATAGA...  1.0  1.0\n",
       "2                AAAACAGTAGTGTACCGCTGTGTATAAATAACCTAAGTGATAAAAG...  1.0  1.0\n",
       "3                AAAACTTAAAGTCCTTTAATAGTATGGCTGAAGCGATTGTTTTCAA...  1.0  1.0\n",
       "4                AAAACTTTTTCTCAAAGTTATAAATCCCTTACTCCCCCCTAAACCT...  1.0  1.0\n",
       "...                                                            ...  ...  ...\n",
       "1556             TTTCCATACTACGTGTACGAAGTTTTATATAATAAATAATAAATTA...  1.0  1.0\n",
       "1557             TTTCTGAAGAATCGATAAGACAAATCAAAACAAATTAACAATCTCA...  1.0  1.0\n",
       "1558             TTTGAATACAACACACATAGACGGCTCTAATTTGAAAAACATAAAA...  1.0  1.0\n",
       "1559             TTTGCACCGGGGATCCTAGGCAAAGTGGTTATCTATTTTGGGACTC...  1.0  1.0\n",
       "1560             TTTGCTTCATTATACTTGAGCAGTGTGTGACATCTGAGGCAATGTT...  1.0  1.0\n",
       "\n",
       "[1561 rows x 3 columns]"
      ]
     },
     "execution_count": 166,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"\"\"\n",
    "2. Compare if duplicates occur between GCA and GCF assemblies of the same virus\n",
    "If they are simply different versions of the same assembly, we may prefer to keep only one.\n",
    "\n",
    "We extract 'assembly_prefix' and 'assembly_id' from the file names.\n",
    "\n",
    "Why Extract the Assembly ID?\n",
    "- Identify if duplicates are between GCA/GCF of the same virus (same assembly_id).\n",
    "- Check if duplicates occur in different viruses (different assembly_id  likely real conservation).\n",
    "- Ensure that if we remove duplicates, we are keeping the best assembly.\n",
    "\"\"\"\n",
    "\n",
    "# Extract file prefixes (e.g., GCA_XXXXX and GCF_XXXXX)\n",
    "# dna_df['assembly_prefix'] = dna_df['file'].str.extract(r'^(GCA|GCF)_\\d+')\n",
    "dna_df[['assembly_pefix', 'assembly_id']] = dna_df['file'].str.extract(r'^(GCA|GCF)_(\\d+\\.\\d+)')\n",
    "# dna_df[['file', 'assembly_prefix', 'assembly_id']].head()\n",
    "\n",
    "# Find duplicates occurring across GCA/GCF pairs (i.e., same sequence in both GCA and GCF)\n",
    "dna_dups = dna_df[dna_df.duplicated(subset=['dna'], keep=False)].sort_values(['dna'])\n",
    "dna_dups = dna_dups[['file', 'assembly_prefix', 'dna']]\n",
    "gca_gcf_dups = dna_dups.groupby(['dna', 'assembly_prefix'])['file'].nunique().unstack()\n",
    "\n",
    "# Display cases where duplicates exist in both GCA and GCF\n",
    "gca_gcf_dups.dropna().reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 167,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>file</th>\n",
       "      <th>quality</th>\n",
       "      <th>id</th>\n",
       "      <th>replicon_type</th>\n",
       "      <th>replicon_geometry</th>\n",
       "      <th>dna</th>\n",
       "      <th>length</th>\n",
       "      <th>assembly_prefix</th>\n",
       "      <th>virus_id</th>\n",
       "      <th>assembly_id</th>\n",
       "      <th>assembly_pefix</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GCA_000847345.1.qual.gto</td>\n",
       "      <td>Good</td>\n",
       "      <td>DQ375403.1</td>\n",
       "      <td>Large RNA Segment</td>\n",
       "      <td>linear</td>\n",
       "      <td>ACACAAAGGCGCCCAATCATGGATTCTATATTATCAAAACAGCTGG...</td>\n",
       "      <td>6404</td>\n",
       "      <td>GCA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>000847345.1</td>\n",
       "      <td>GCA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GCA_000847345.1.qual.gto</td>\n",
       "      <td>Good</td>\n",
       "      <td>DQ380206.1</td>\n",
       "      <td>Medium RNA Segment</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ACACAAAGACGGTGCATTAAATGTATGTTTTATTAACAATTCTAAT...</td>\n",
       "      <td>3885</td>\n",
       "      <td>GCA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>000847345.1</td>\n",
       "      <td>GCA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GCA_000847345.1.qual.gto</td>\n",
       "      <td>Good</td>\n",
       "      <td>DQ380151.1</td>\n",
       "      <td>Small RNA Segment</td>\n",
       "      <td>linear</td>\n",
       "      <td>ACACAAAGACCCCCTAGTGCTTATCAAGTATATCATGGATTACTTT...</td>\n",
       "      <td>1690</td>\n",
       "      <td>GCA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>000847345.1</td>\n",
       "      <td>GCA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GCA_000847645.1.qual.gto</td>\n",
       "      <td>Good</td>\n",
       "      <td>AB009656.1</td>\n",
       "      <td>Large RNA Segment</td>\n",
       "      <td>linear</td>\n",
       "      <td>ACACAAAGTCCTGGACAACAAAAACAAAAAAACTCTTTCATCAATA...</td>\n",
       "      <td>9760</td>\n",
       "      <td>GCA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>000847645.1</td>\n",
       "      <td>GCA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GCA_000847645.1.qual.gto</td>\n",
       "      <td>Good</td>\n",
       "      <td>AB010376.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ACACAAAGTCCTGGGCAAAAGAAAAGCAAAATAACCAACTATAGAT...</td>\n",
       "      <td>4056</td>\n",
       "      <td>GCA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>000847645.1</td>\n",
       "      <td>GCA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4681</th>\n",
       "      <td>GCF_029888565.1.qual.gto</td>\n",
       "      <td>Good</td>\n",
       "      <td>NC_079116.1</td>\n",
       "      <td>Medium RNA Segment</td>\n",
       "      <td>linear</td>\n",
       "      <td>AGTAGTGTACTACCAATTATACAAAAGTTTCATTCAAAATGTTGTT...</td>\n",
       "      <td>4351</td>\n",
       "      <td>GCF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>029888565.1</td>\n",
       "      <td>GCF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4682</th>\n",
       "      <td>GCF_029888565.1.qual.gto</td>\n",
       "      <td>Good</td>\n",
       "      <td>NC_079118.1</td>\n",
       "      <td>Small RNA Segment</td>\n",
       "      <td>linear</td>\n",
       "      <td>AGTAGTGTACTCCACAAAAAACAAAACTATAAATCGTTAGAAACAA...</td>\n",
       "      <td>923</td>\n",
       "      <td>GCF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>029888565.1</td>\n",
       "      <td>GCF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4683</th>\n",
       "      <td>GCF_031497195.1.qual.gto</td>\n",
       "      <td>Good</td>\n",
       "      <td>NC_086346.1</td>\n",
       "      <td>Small RNA Segment</td>\n",
       "      <td>linear</td>\n",
       "      <td>ACACAAAGTTACCTAAACTTTGTGTTTGAGTCCTGAAATACAAACA...</td>\n",
       "      <td>1863</td>\n",
       "      <td>GCF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>031497195.1</td>\n",
       "      <td>GCF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4684</th>\n",
       "      <td>GCF_031497195.1.qual.gto</td>\n",
       "      <td>Good</td>\n",
       "      <td>NC_086347.1</td>\n",
       "      <td>Medium RNA Segment</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CTTTGTGTTTGACGGTGTTAATGAATTCAAAAAACATTTTTTATCT...</td>\n",
       "      <td>4310</td>\n",
       "      <td>GCF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>031497195.1</td>\n",
       "      <td>GCF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4685</th>\n",
       "      <td>GCF_031497195.1.qual.gto</td>\n",
       "      <td>Good</td>\n",
       "      <td>NC_086348.1</td>\n",
       "      <td>Large RNA Segment</td>\n",
       "      <td>linear</td>\n",
       "      <td>ACACAAAGGCGCCCAATCATGGATGACATAATATCAAAACAAGTTG...</td>\n",
       "      <td>6374</td>\n",
       "      <td>GCF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>031497195.1</td>\n",
       "      <td>GCF</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4686 rows  11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                          file quality           id       replicon_type  \\\n",
       "0     GCA_000847345.1.qual.gto    Good   DQ375403.1   Large RNA Segment   \n",
       "1     GCA_000847345.1.qual.gto    Good   DQ380206.1  Medium RNA Segment   \n",
       "2     GCA_000847345.1.qual.gto    Good   DQ380151.1   Small RNA Segment   \n",
       "3     GCA_000847645.1.qual.gto    Good   AB009656.1   Large RNA Segment   \n",
       "4     GCA_000847645.1.qual.gto    Good   AB010376.1                 NaN   \n",
       "...                        ...     ...          ...                 ...   \n",
       "4681  GCF_029888565.1.qual.gto    Good  NC_079116.1  Medium RNA Segment   \n",
       "4682  GCF_029888565.1.qual.gto    Good  NC_079118.1   Small RNA Segment   \n",
       "4683  GCF_031497195.1.qual.gto    Good  NC_086346.1   Small RNA Segment   \n",
       "4684  GCF_031497195.1.qual.gto    Good  NC_086347.1  Medium RNA Segment   \n",
       "4685  GCF_031497195.1.qual.gto    Good  NC_086348.1   Large RNA Segment   \n",
       "\n",
       "     replicon_geometry                                                dna  \\\n",
       "0               linear  ACACAAAGGCGCCCAATCATGGATTCTATATTATCAAAACAGCTGG...   \n",
       "1                  NaN  ACACAAAGACGGTGCATTAAATGTATGTTTTATTAACAATTCTAAT...   \n",
       "2               linear  ACACAAAGACCCCCTAGTGCTTATCAAGTATATCATGGATTACTTT...   \n",
       "3               linear  ACACAAAGTCCTGGACAACAAAAACAAAAAAACTCTTTCATCAATA...   \n",
       "4                  NaN  ACACAAAGTCCTGGGCAAAAGAAAAGCAAAATAACCAACTATAGAT...   \n",
       "...                ...                                                ...   \n",
       "4681            linear  AGTAGTGTACTACCAATTATACAAAAGTTTCATTCAAAATGTTGTT...   \n",
       "4682            linear  AGTAGTGTACTCCACAAAAAACAAAACTATAAATCGTTAGAAACAA...   \n",
       "4683            linear  ACACAAAGTTACCTAAACTTTGTGTTTGAGTCCTGAAATACAAACA...   \n",
       "4684               NaN  CTTTGTGTTTGACGGTGTTAATGAATTCAAAAAACATTTTTTATCT...   \n",
       "4685            linear  ACACAAAGGCGCCCAATCATGGATGACATAATATCAAAACAAGTTG...   \n",
       "\n",
       "      length assembly_prefix virus_id  assembly_id assembly_pefix  \n",
       "0       6404             GCA      NaN  000847345.1            GCA  \n",
       "1       3885             GCA      NaN  000847345.1            GCA  \n",
       "2       1690             GCA      NaN  000847345.1            GCA  \n",
       "3       9760             GCA      NaN  000847645.1            GCA  \n",
       "4       4056             GCA      NaN  000847645.1            GCA  \n",
       "...      ...             ...      ...          ...            ...  \n",
       "4681    4351             GCF      NaN  029888565.1            GCF  \n",
       "4682     923             GCF      NaN  029888565.1            GCF  \n",
       "4683    1863             GCF      NaN  031497195.1            GCF  \n",
       "4684    4310             GCF      NaN  031497195.1            GCF  \n",
       "4685    6374             GCF      NaN  031497195.1            GCF  \n",
       "\n",
       "[4686 rows x 11 columns]"
      ]
     },
     "execution_count": 167,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dna_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 161,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>file</th>\n",
       "      <th>assembly_prefix</th>\n",
       "      <th>dna</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2172</th>\n",
       "      <td>GCA_018595275.1.qual.gto</td>\n",
       "      <td>GCA</td>\n",
       "      <td>AAAAAAAAAAAAAAAAAAAAAAAAAAAGTAGTAAACACACTCATAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4391</th>\n",
       "      <td>GCF_018595275.1.qual.gto</td>\n",
       "      <td>GCF</td>\n",
       "      <td>AAAAAAAAAAAAAAAAAAAAAAAAAAAGTAGTAAACACACTCATAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2467</th>\n",
       "      <td>GCA_029888265.1.qual.gto</td>\n",
       "      <td>GCA</td>\n",
       "      <td>AAAACAAAAAGAGATGGATCAAGACGTAGATCAAATTCTAAATAGA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4647</th>\n",
       "      <td>GCF_029888265.1.qual.gto</td>\n",
       "      <td>GCF</td>\n",
       "      <td>AAAACAAAAAGAGATGGATCAAGACGTAGATCAAATTCTAAATAGA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3710</th>\n",
       "      <td>GCF_002831345.1.qual.gto</td>\n",
       "      <td>GCF</td>\n",
       "      <td>AAAACAGTAGTGTACCGCTGTGTATAAATAACCTAAGTGATAAAAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>355</th>\n",
       "      <td>GCA_001924595.1.qual.gto</td>\n",
       "      <td>GCA</td>\n",
       "      <td>TTTGAATACAACACACATAGACGGCTCTAATTTGAAAAACATAAAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>215</th>\n",
       "      <td>GCA_000930735.1.qual.gto</td>\n",
       "      <td>GCA</td>\n",
       "      <td>TTTGCACCGGGGATCCTAGGCAAAGTGGTTATCTATTTTGGGACTC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3333</th>\n",
       "      <td>GCF_000930735.1.qual.gto</td>\n",
       "      <td>GCF</td>\n",
       "      <td>TTTGCACCGGGGATCCTAGGCAAAGTGGTTATCTATTTTGGGACTC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4324</th>\n",
       "      <td>GCF_018594925.1.qual.gto</td>\n",
       "      <td>GCF</td>\n",
       "      <td>TTTGCTTCATTATACTTGAGCAGTGTGTGACATCTGAGGCAATGTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2105</th>\n",
       "      <td>GCA_018594925.1.qual.gto</td>\n",
       "      <td>GCA</td>\n",
       "      <td>TTTGCTTCATTATACTTGAGCAGTGTGTGACATCTGAGGCAATGTT...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3208 rows  3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                          file assembly_prefix  \\\n",
       "2172  GCA_018595275.1.qual.gto             GCA   \n",
       "4391  GCF_018595275.1.qual.gto             GCF   \n",
       "2467  GCA_029888265.1.qual.gto             GCA   \n",
       "4647  GCF_029888265.1.qual.gto             GCF   \n",
       "3710  GCF_002831345.1.qual.gto             GCF   \n",
       "...                        ...             ...   \n",
       "355   GCA_001924595.1.qual.gto             GCA   \n",
       "215   GCA_000930735.1.qual.gto             GCA   \n",
       "3333  GCF_000930735.1.qual.gto             GCF   \n",
       "4324  GCF_018594925.1.qual.gto             GCF   \n",
       "2105  GCA_018594925.1.qual.gto             GCA   \n",
       "\n",
       "                                                    dna  \n",
       "2172  AAAAAAAAAAAAAAAAAAAAAAAAAAAGTAGTAAACACACTCATAT...  \n",
       "4391  AAAAAAAAAAAAAAAAAAAAAAAAAAAGTAGTAAACACACTCATAT...  \n",
       "2467  AAAACAAAAAGAGATGGATCAAGACGTAGATCAAATTCTAAATAGA...  \n",
       "4647  AAAACAAAAAGAGATGGATCAAGACGTAGATCAAATTCTAAATAGA...  \n",
       "3710  AAAACAGTAGTGTACCGCTGTGTATAAATAACCTAAGTGATAAAAG...  \n",
       "...                                                 ...  \n",
       "355   TTTGAATACAACACACATAGACGGCTCTAATTTGAAAAACATAAAA...  \n",
       "215   TTTGCACCGGGGATCCTAGGCAAAGTGGTTATCTATTTTGGGACTC...  \n",
       "3333  TTTGCACCGGGGATCCTAGGCAAAGTGGTTATCTATTTTGGGACTC...  \n",
       "4324  TTTGCTTCATTATACTTGAGCAGTGTGTGACATCTGAGGCAATGTT...  \n",
       "2105  TTTGCTTCATTATACTTGAGCAGTGTGTGACATCTGAGGCAATGTT...  \n",
       "\n",
       "[3208 rows x 3 columns]"
      ]
     },
     "execution_count": 161,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dna_dups"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 158,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # Explore commands\n",
    "# dna_dups.groupby(['dna', 'assembly_prefix'])['file'].nunique().reset_index()\n",
    "# dna_dups.groupby(['dna', 'assembly_prefix'])['file'].nunique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "num_unique_viruses\n",
       "0    1598\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 91,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"\"\"\n",
    "3. Check if duplicates occur across different viruses\n",
    "This helps determine if segments are widely shared or just repeated within the same virus.\n",
    "\"\"\"\n",
    "\n",
    "# Extract virus ID from 'file' (assuming it's part of the filename)\n",
    "dna_df['virus_id'] = dna_df['file'].str.extract(r'G[CA]F?_(\\d+)\\.qual')\n",
    "\n",
    "# Count how many unique viruses each duplicate appears in\n",
    "dup_virus_counts = dna_df[dna_df.duplicated(subset=['dna'], keep=False)] \\\n",
    "    .groupby('dna')['virus_id'].nunique().reset_index()\n",
    "\n",
    "dup_virus_counts.rename(columns={'virus_id': 'num_unique_viruses'}, inplace=True)\n",
    "dup_virus_counts['num_unique_viruses'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "4. Identify the best sequence to keep\n",
    "For each set of duplicates, prioritize sequences as follows:\n",
    "- Prefer Good quality over Poor.\n",
    "- Prefer GCF (reference assembly) over GCA.\n",
    "- If both are the same, keep only one per virus to avoid redundancy.\n",
    "\"\"\"\n",
    "\n",
    "# Sort by quality and assembly type to prioritize what to keep\n",
    "dna_df['quality_score'] = dna_df['quality'].map({'Good': 1, 'Poor': 0})\n",
    "dna_df['assembly_score'] = dna_df['file'].str.startswith('GCF').astype(int)\n",
    "\n",
    "# Rank duplicates within each sequence group\n",
    "dna_df['dup_rank'] = dna_df.groupby('dna')[['quality_score', 'assembly_score']].rank(\n",
    "    method='first', ascending=False).sum(axis=1)\n",
    "\n",
    "# Mark sequences to keep (rank = 1 means best sequence)\n",
    "dna_df['keep'] = dna_df['dup_rank'] == 1\n",
    "\n",
    "# Show cases where we would remove some duplicates\n",
    "dna_df[dna_df.duplicated(subset=['dna'], keep=False) & ~dna_df['keep']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "display(df[:20])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "display(df['quality'].value_counts())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df['type'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df['function'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "cepi",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.21"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
